dex/networks/eth/contractor.go

```diff
@@ -0,0 +1,192 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+//go:build lgpl
+// +build lgpl
+
+package eth
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/encode"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// Contractor is a translation layer between the abigen bindings and the DEX app.
+// The intention is that if a new contract is implemented, the Contractor
+// interface itself will not require any updates.
+type Contractor interface {
+	Swap(ctx context.Context, secretHash [32]byte) (*SwapState, error)
+	Initiate(*bind.TransactOpts, []*asset.Contract) (*types.Transaction, error)
+	Redeem(txOpts *bind.TransactOpts, redeems []*asset.Redemption) (*types.Transaction, error)
+	Refund(opts *bind.TransactOpts, secretHash [32]byte) (*types.Transaction, error)
+	EstimateInitGas(ctx context.Context, n int) (uint64, error)
+	EstimateRedeemGas(ctx context.Context, secrets [][32]byte) (uint64, error)
+	IsRedeemable(secretHash, secret [32]byte) (bool, error)
+}
+
+var ContractAddresses = map[uint32]map[dex.Network]common.Address{
+	0: {
+		dex.Mainnet: common.Address{},
+		dex.Simnet:  common.Address{},
+		dex.Testnet: common.Address{},
+	},
+}
+
+type ContractorConstructor func(net dex.Network, addr common.Address, ec *ethclient.Client) (Contractor, error)
+
+// contractorv0 is the Contractor for contract version 0.
+// Redeem and Refund methods of ETHSwap already have suitable return types.
+type contractorV0 struct {
+	*ETHSwap
```

Comment:
We should save the following suggestion for another day, but since `ETHSwap` is the v0 contract API compiled from it's solidity, we'll probably want to move contract.go -> contracts/v0.go or contracts/apis/v0.go